### ISSUE TITLE
Almost perfect Baarle-Nassau fit with 1000-wide model

### ISSUE THREAD

**Issue by inikishev** [WelchLabs/videos#14]
Hello!

So apparently 1000 neurons is enough to fit the Baarle-Nassau borders, I was able to achieve 99.96% accuracy using Shor's r-algorithm as the optimizer:

<img width="518" height="511" alt="Image" src="https://github.com/user-attachments/assets/db39c8fc-5152-46ef-97a7-ada40495d50b" />

here is the notebook https://github.com/inikishev/inikishev/blob/main/notebooks/Baarle-Hertog%20border%20fitting/1k%20ShorR%20large%20batch%201000%20epochs.ipynb

I changed Adam to Shor-s r-algorithm and batch size from 10,000 to 100,000 and ran it for 1000 epochs.

Shor-R is a method for non-smooth optimization so it works well for shallow ReLU-nets. It is related to the conjugate gradient method which prevents zig-zagging by making each next update direction orthogonal to all differences between past consecutive gradients, in Shor-R it is not made completely orthogonal but rather dilated by a factor determined by a hyperparameter. It does require a N^2 memory to store the dilation matrix so it is only feasible for problems with under around 10,000 parameters. Also since it uses differences between consecutive gradients, it is not really suitable for mini-batch training, but even batch size of 1,000 appears to work fine (that way I got 0.9964 accuracy in 20 epochs, and 0.9155 for 32-wide model). Also it requires a line search so I used strong-wofle.

A 500-wide model reached 99.74 accuracy

<img width="518" height="511" alt="Image" src="https://github.com/user-attachments/assets/4b20f06b-06ed-40f3-ad73-9ceff8d5a4f9" />

100-wide model reached 96.63%

<img width="518" height="511" alt="Image" src="https://github.com/user-attachments/assets/b5356f67-6d8a-4c8d-bfea-14464727d2a9" />

 32-wide model reached 91.55%

<img width="518" height="511" alt="Image" src="https://github.com/user-attachments/assets/c4eb32ce-58cc-4d62-b64f-9c38c1ea25a8" />

**Comment by stephencwelch** [comment]
Woah this is so cool! Will review when I get a little time!
